### PR TITLE
Datamodel init tool

### PR DESCRIPTION
**Pull request by olivierdalang**
Hi !

As quickly presented during last QGEP day, this is the current state of the QGEP data model initialisation tool.

Features :
- downloads the datamodel automatically
- one button to install python dependencies
- pg_services.conf selector/editor
- can initialize new datamodels (empty structure)
- can upgrade existing datamodels
- can load the template QGIS project (adapted to work with the selected pg_service)

We've tested it with @urskaufmann and it seems it's now in a more or less usable state. It would be nice to get a bit more testing on different setup and workflows before merging. Currently it's been tested on Win 7/10, both with admin account and in a more secured enterprise setting. More testing on Windows and some testing on Linux would be nice.

Some features are planned but not implemented yet :

- [ ] Add ability to specify OID prefix on database initialization
- [ ] Add the ability to store a custom QGIS template project instead of the default one
- [ ] Allow for password prompt if credential are required and absent from pg_service.conf
- [ ] Add some documentation
- [x] Deal with the fact that we may not want to expose this tool to all users. Any suggestion about the best way to deal with that would be very welcome, as at the same time, we don't want to hide it too much (for instance requiring a command to be run in the python console to enable it may be a bit too hidden, as we are targeting users that don't like the command line too much)

### CODE REVIEW

**Comment by m-kuhn** [comment]
@olivierdalang feel free to silence warnings (E501 foremost)

**Comment by olivierdalang** [comment]
(putting this on hold for now until pum2 strategy/planned date is more clear)

**Comment by olivierdalang** [comment]
As per discussion 03/06/2021 : OK to merge this, but add a hidden setting to hide it by default for users

**Comment by olivierdalang** [comment]
So I added a setting (defaults to false) that defines whether the datainit tool appears in the menu.
It's accessible in the UI settings, yet not too proeminent.

Is that OK ? If so I think this is ready to review/merge.

image

**Comment by olivierdalang** [comment]
@pep8speaks pep8ify

File: qgepplugin/gui/qgepdatamodeldialog.py [hunk 1]
```diff
@@ -0,0 +1,773 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------
+#
+# Profile
+# Copyright (C) 2012  Matthias Kuhn
```
Comment (m-kuhn) [inline qgepplugin/gui/qgepdatamodeldialog.py:5]: Sure about that?

File: qgepplugin/gui/qgepdatamodeldialog.py [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+from qgis.PyQt.QtNetwork import QNetworkRequest, QNetworkReply
+
+from qgis.core import QgsMessageLog, QgsNetworkAccessManager, Qgis, QgsProject
+
+from ..utils import get_ui_class
+
+# TODO : get latest dynamically ?
+AVAILABLE_RELEASES = {
+    # 'master': 'https://github.com/QGEP/datamodel/archive/master.zip',  # TODO : if we expose this here, we should put a big red warning and not take it default
+    '1.5.3': 'https://github.com/QGEP/datamodel/archive/1.5.3.zip',
```
Comment (m-kuhn) [inline qgepplugin/gui/qgepdatamodeldialog.py:47]:
```suggestion
    '1.5.4': 'https://github.com/QGEP/datamodel/archive/1.5.4.zip',
```

I guess that proves the point of line 44

File: qgepplugin/gui/qgepdatamodeldialog.py [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+    def _show_error(self, message):
+        self._done_progress()
+        err = QMessageBox()
+        err.setText(message)
+        err.setIcon(QMessageBox.Warning)
+        err.exec_()
+
+    # Actions helpers
+
+    def _run_sql(self, connection_string, sql_command, autocommit=False, error_message='Psycopg error, see logs for more information'):
```
[the remark this answers is not in the document]
Comment (olivierdalang) [reply qgepplugin/gui/qgepdatamodeldialog.py:231]: Indeed it's not really a gui thing, but at the same time it's very specific to that dialog, and not sure we want to promote using this elsewhere (hence prefixing with `_` to mark as private) 
As discussed with Stefan, Arnaud and Urs, this whole datamodel init tool is not really a long term solution as we should rather move this type of functionality to pum (2).
So IMO still makes sense to keep this file self-contained (anyway we can spaghettify the code the day we need that from other place of the code)

File: qgepplugin/gui/qgepdatamodeldialog.py (hunk 1 above)
Comment (olivierdalang) [reply qgepplugin/gui/qgepdatamodeldialog.py:5]: Ok will change (I never understand what these headers a good for, also in that case the origin is probably indeed your file from a copy/paste ?)

File: qgepplugin/gui/qgepdatamodeldialog.py (hunk 2 above)
Comment (olivierdalang) [reply qgepplugin/gui/qgepdatamodeldialog.py]: Added a note about that :
```
# Currently, the latest release is hard-coded in the plugin, meaning we need
# to publish a plugin update for each datamodel update.
# In the future, once plugin/datamodel versionning scheme clearly reflects
# compatibility, we could retrieve this dynamically, so datamodel bugfix
# releases don't require a plugin upgrade.
```

What do you think ?

**Comment by olivierdalang** [comment]
(made a super-mess with the reformatting, we'll squash this before merging)

File: qgepplugin/gui/qgepdatamodeldialog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-kuhn) [reply qgepplugin/gui/qgepdatamodeldialog.py]: Agreed. But I prefer lasagne.

**Pull request merged**